Entropy estimation with non-invertible generative models

austin-hoover opened the pull request:
This PR adds the class `NNDist` to the `gpsr.beams` module. With a non-invertible transformation from the base distribution (`NNTransform`), we can't evaluate the probability density $p(x)$ at an arbitrary point $x$. But we can still evaluate the density at a batch of sampled points $\\{ x^{(i)} \\} \sim p(x)$ by computing the Jacobian matrix of the transformation. This can provide differentiable estimates of the KL divergence or other integrals involving the distribution function.


## Changes

* Added method `compute_batched_jacobian(x: torch.Tensor, function: Callable) -> torch.Tensor:` to `gpsr.beams` module. This method computes the Jacobian matrix of callable `function` for a batch of particles `x`, following the example [here](https://docs.pytorch.org/functorch/stable/generated/functorch.jacrev.html).
* Added class `gpsr.beams.NNDist`. This is a subclass of `GenModel` that uses the `NNTransform` as the transformation from the base distribution. It has methods `sample` and `sample_and_log_prob` but not `log_prob`.
* Renamed class `gpsr.beams.NSF` to `gpsr.beams.NSFDist` for consistency. Also changed arguments in `NSFDist` to match `NNDist` (`width`, `depth`).
* Added method `entropy(n: int, prior: Any) -> torch.Tensor` to the `gpsr.beams.GenModel` class. This is a convenience method to estimate the entropy or KL divergence from a batch of `n` samples, using the `sample_and_log_prob` method.
* Added [examples](https://github.com/austin-hoover/gpsr/tree/jacobian/docs/examples/kl) which train the model by minimizing the KL divergence.

<img height="200" alt="test" src="https://github.com/user-attachments/assets/eba252c0-7240-49b9-9127-bcc6a1e7aa0d" />

austin-hoover committed:
Test jacobian calculation

austin-hoover committed:
Add 6D example

austin-hoover committed:
Modify example

austin-hoover committed:
Merge branch 'roussel-ryan:main' into jacobian

austin-hoover committed:
Change name from NSF to NSFDist

austin-hoover committed:
Update generative model classes

- Add NNDist class
- Default cov_matrix to identity matrix
- Add entropy method to GenModel class

austin-hoover committed:
Change example to use gpsr.beams module

austin-hoover committed:
Format

austin-hoover committed:
Format

austin-hoover committed:
Format

roussel-ryan commented on docs/examples/jacobian/test_jacobian.ipynb:
Can we move this to a unit test instead of having a python notebook?

austin-hoover replied:
I added methods to `test_beams.py` here: https://github.com/austin-hoover/gpsr/blob/913c247477011841ba6d1bf3a47c162b9956a63a/gpsr/tests/test_beams.py#L40-L71. The first one tests that Jacobian of `torch.sin` is correct. The second one computes the Jacobian of the `NNTransform` forward method.

roussel-ryan commented on docs/examples/jacobian/test_kl_nn.ipynb:
this is closer to a usable example, so we should rename to kl_nn.ipynb. A separate unit test would be useful for validating changes to beams.py

austin-hoover replied:
Renamed kl_nn.ipynb

roussel-ryan reviewed (commented):
Really cool stuff @austin-hoover . Is there a reference for the batched jacobian method of estimating the entropy using this technique?

austin-hoover committed:
Change name from test_kl_nn to test_kl.ipynb

austin-hoover committed:
Add test test_jacobian_sin

austin-hoover committed:
Delete test_jacobian.ipynb

austin-hoover committed:
Add test test_nn_transform_jacobian

austin-hoover commented:
For references: the formula for KL divergence is the same as for normalizing flows (Section 2.3.2 of https://arxiv.org/pdf/2404.00502). There's some discussion about computing Jacobian determinant for non-invertible models [here](https://arxiv.org/pdf/2404.00502).

austin-hoover committed:
Format

austin-hoover committed:
pylint

austin-hoover committed:
Rename arguments in NSFDist to match NNDist

austin-hoover committed:
Update Jacobian examples

austin-hoover committed:
Format

austin-hoover committed:
Rename examples/jacobian to examples/kl

austin-hoover committed:
Format

austin-hoover committed:
Fix log_prob

x = Mz
p(x) = p(z) / det(M) = p(z) * det(inv(M))

Only affects log_prob function; sample_and_log_prob was correct.

austin-hoover committed:
Reverts 8a3ea6aa0391ad761b40ebf01e38aba3727bb787

Confused myself.

austin-hoover committed:
Rename variables

austin-hoover committed:
Add reverse KL minimization examples with ring distribution

Here it is easier to see when the model is failing

austin-hoover committed:
Merge branch 'roussel-ryan:main' into jacobian

austin-hoover commented:
Marking this as a draft for now.

* The KL minimization looks like it's working in 2D examples, but convergence is slower than the normalizing flow and sometimes fails to match the target distribution, depending on initialization. So far I can't reproduce correct MaxEnt solutions for 2D reconstruction.
* I also tried to train the NN using a sliced Wasserstein loss function to regress to a target distribution. In this case, the NN has an even harder time converging. So maybe the log-prob calculation is fine and these issues have to do with the specific NN architecture used.

austin-hoover committed:
Merge branch 'roussel-ryan:main' into jacobian

austin-hoover committed:
Shorten comment

austin-hoover committed:
Add test of log_prob calculation using NSF

austin-hoover committed:
Merge branch 'jacobian' of https://github.com/austin-hoover/gpsr into jacobian

austin-hoover committed:
Update kl examples

austin-hoover commented:
Added test under /gpsr/tests/test_beams.py that verifies the log_prob calculation is the same as the exact result from the normalizing flow.

roussel-ryan commented:
Thanks @austin-hoover. I'm looking into the testing errors now since its effecting the main branch
